ldap/servers/plugins/acl/acl.c

```diff
@@ -2773,64 +2743,35 @@ acl__resource_match_aci(Acl_PBlock *aclpb, aci_t *aci, int skip_attrEval, int *a
                 matches = (attr_matched ? ACL_TRUE : ACL_FALSE);
             }
 
-
             aclpb->aclpb_state &= ~ACLPB_ATTR_STAR_MATCHED;
             /* figure out how it matched, i.e star matched */
-            if (matches && star_matched && num_attrs == 1 &&
-                !(aclpb->aclpb_state & ACLPB_FOUND_ATTR_RULE))
+            if (matches && star_matched && num_attrs == 1 && !(aclpb->aclpb_state & ACLPB_FOUND_ATTR_RULE)) {
                 aclpb->aclpb_state |= ACLPB_ATTR_STAR_MATCHED;
-            else {
+            } else {
                 /* we are here means that there is a specific
-                ** attr in the rule for this resource.
-                ** We need to avoid this case
-                ** Rule 1: (targetattr = "uid")
-                ** Rule 2: (targetattr = "*")
-                ** we cannot use STAR optimization
-                */
+                 * attr in the rule for this resource.
+                 * We need to avoid this case
+                 * Rule 1: (targetattr = "uid")
+                 * Rule 2: (targetattr = "*")
+                 * we cannot use STAR optimization
+                 */
                 aclpb->aclpb_state |= ACLPB_FOUND_ATTR_RULE;
                 aclpb->aclpb_state &= ~ACLPB_ATTR_STAR_MATCHED;
             }
-        } else if ((c_attrEval) ||
-                   (aci->aci_type & ACI_TARGET_ATTR)) {
-            if ((aci_right & ACL_RIGHTS_TARGETATTR_NOT_NEEDED) &&
-                (aclpb->aclpb_access & ACL_RIGHTS_TARGETATTR_NOT_NEEDED)) {
-                /*
-            ** Targetattr rule doesn't  make any sense
-            ** in this case. So select this rule
-            ** default: matches = ACL_TRUE;
-            */
-                ;
-            } else if (aci_right & SLAPI_ACL_WRITE &&
-                       (aci->aci_type & ACI_TARGET_ATTR) &&
-                       !(c_attrEval) &&
-                       (aci->aci_type & ACI_HAS_ALLOW_RULE)) {
-                /* We need to handle modrdn operation.  Modrdn doesn't
-            ** change any attrs but changes the RDN and so (attr=NULL).
-            ** Here we found an acl which has a targetattr but
-            ** the resource doesn't need one. In that case, we should
-            ** consider this acl.
-            ** the opposite is true if it is a deny rule, only a deny without
-            ** any targetattr should deny modrdn
-            ** default: matches = ACL_TRUE;
-            */
-                ;
-            } else {
-                matches = ACL_FALSE;
-            }
         }
```

Comment:
This code removal above is the "fix"...

Reply:
The removal of this code does not cause any regressions in the ACI test suite, but I am not sure it is 100% valid either because the purpose of this code block is not entirely clear.   Please review...

Reply:
Doing some git investigation (and having to checkout from before the great code format clean up of 2017), I can see the following in git blame:

```
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2790)    } else if ( (c_attrEval) ||
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2791)                (aci->aci_type & ACI_TARGET_ATTR)) {
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2792)            if ((aci_right & ACL_RIGHTS_TARGETATTR_NOT_NEEDED) &&
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2793)                (aclpb->aclpb_access & ACL_RIGHTS_TARGETATTR_NOT_NEEDED)) {
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2794)                    /*
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2795)                    ** Targetattr rule doesn't  make any sense
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2796)                    ** in this case. So select this rule
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2797)                    ** default: matches = ACL_TRUE;
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2798)                    */
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2799)                    ;
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2800)            } else if (aci_right & SLAPI_ACL_WRITE &&
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2801)                      (aci->aci_type & ACI_TARGET_ATTR) &&
fe0491c49f (Ludwig Krispenz          2013-07-09 16:11:12 +0200 2802)                      !(c_attrEval) &&
fe0491c49f (Ludwig Krispenz          2013-07-09 16:11:12 +0200 2803)                      (aci->aci_type & ACI_HAS_ALLOW_RULE)) {
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2804)                    /* We need to handle modrdn operation.  Modrdn doesn't
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2805)                    ** change any attrs but changes the RDN and so (attr=NULL).
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2806)                    ** Here we found an acl which has a targetattr but
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2807)                    ** the resource doesn't need one. In that case, we should
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2808)                    ** consider this acl.
fe0491c49f (Ludwig Krispenz          2013-07-09 16:11:12 +0200 2809)                    ** the opposite is true if it is a deny rule, only a deny without
fe0491c49f (Ludwig Krispenz          2013-07-09 16:11:12 +0200 2810)                    ** any targetattr should deny modrdn
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2811)                    ** default: matches = ACL_TRUE;
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2812)                    */
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2813)                    ;

```

It looks like the majority of this code is "quite old" to say the least, but what is interesting is the condition change from ludwig in 2013. According to git its:

```
commit fe0491c49fdb180f1bd291ba78148da7c58b5e90
Author: Ludwig Krispenz <lkrispen@redhat.com>
Date:   Tue Jul 9 16:11:12 2013 +0200

    Ticket 47399 - RHDS denies MODRDN access if ACI list contains any DENY rule

    Bug Description:  if there is a deny rule targeting a specific attribute it also
                      denies the modrdn operation

    Fix Description:   only apply deny rules to modrdn if no target attr exists

    https://fedorahosted.org/389/ticket/47339

    Reviewed by: Rich, thanks
```

So I'm not sure the delete of this block is the right change here, especialy given how ... fun the aci code is at the best of times. I think we need some more research?